lib/classroom_api.rb

```diff
@@ -8,190 +8,229 @@ def initialize(access_token)
-        @log.api_logger.debug "add_facility_id_to_classrooms, error: could not delete records with #{@rooms_in_db} rmrecnbr"
+        @log.api_logger.debug "add_facility_id_to_classrooms, error: API return: #{result['errorcode']} - #{result['error']} for #{facility_id}"
         @debug = true
         return @debug
       end
+      # check if database has rooms that are not in API anymore
+      if @rooms_in_db.present?
+        RoomContact.where(rmrecnbr: @rooms_in_db).delete_all
+        RoomCharacteristic.where(rmrecnbr: @rooms_in_db).delete_all
+        if Room.where(rmrecnbr: @rooms_in_db).delete_all
+          @log.api_logger.info "add_facility_id_to_classrooms, delete #{@rooms_in_db} room(s) from the database"
+        else
+          @log.api_logger.debug "add_facility_id_to_classrooms, error: could not delete records with #{@rooms_in_db} rmrecnbr"
+          @debug = true
+          return @debug
+        end
+      end
+    rescue StandardError => e
+      # example: Errno::ETIMEDOUT: Operation timed out - user specified timeout
+      @log.api_logger.debug "add_facility_id_to_classrooms, error: API return: #{e.message}"
+      @debug = true
     end
     return @debug
   end
 
 
   def get_classrooms_list
-    result = {'success' => false, 'errorcode' => '', 'error' => '', 'data' => {}}
-    url = URI("https://gw.api.it.umich.edu/um/aa/ClassroomList/Classrooms?BuildingID=1005046")
-
-    http = Net::HTTP.new(url.host, url.port)
-    http.use_ssl = true
-    http.verify_mode = OpenSSL::SSL::VERIFY_NONE
-
-    request = Net::HTTP::Get.new(url)
-    request["x-ibm-client-id"] = "#{Rails.application.credentials.um_api[:buildings_client_id]}"
-    request["authorization"] = "Bearer #{@access_token}"
-    request["accept"] = 'application/json'
-
-    response = http.request(request)
-    response_json = JSON.parse(response.read_body)
-    if response_json['errorCode'].present?
-      result['errorcode'] = response_json['errorCode']
-      result['error'] = response_json['errorMessage']
-    else
-      result['success'] = true
-      result['data'] = response_json['Classrooms']['Classroom']
+    begin
+      result = {'success' => false, 'errorcode' => '', 'error' => '', 'data' => {}}
+      classrooms = []
+      start_index = 0
+      count = 1000
+      next_page = true
+      while next_page do
+        url = URI("https://gw.api.it.umich.edu/um/aa/ClassroomList/v2/Classrooms?$start_index=#{start_index}&$count=#{count}")
+
+        http = Net::HTTP.new(url.host, url.port)
+        http.use_ssl = true
+        http.verify_mode = OpenSSL::SSL::VERIFY_NONE
+
+        request = Net::HTTP::Get.new(url)
+        request["x-ibm-client-id"] = "#{Rails.application.credentials.um_api[:buildings_client_id]}"
+        request["authorization"] = "Bearer #{@access_token}"
+        request["accept"] = 'application/json'
+
+        response = http.request(request)
+        link = response.to_hash["link"].to_s
+        if link.include? "rel=next"
```

Comment:
The response.to_hash that is returned looks like this:

`{"cache-control"=>["private"], "content-language"=>["en"], "vary"=>["accept-encoding"], "content-type"=>["application/json;subtype=denodo-8.0;charset=UTF-8"], "date"=>["Tue, 26 Mar 2024 02:41:15 GMT"], "link"=>["<https://gw.api.it.umich.edu/um/aa/ClassroomList/v2/Classrooms?$start_index=0&$count=1000>; rel=self, <https://gw.api.it.umich.edu/um/aa/ClassroomList/v2/Classrooms?$count=1000>; rel=prev, <https://gw.api.it.umich.edu/um/aa/ClassroomList/v2/Classrooms?$start_index=1000&$count=1000>; rel=next"], "x-request-id"=>["35c06169-f33f-4161-a4dd-aa45c00cbd5b"], "via"=>["1.1 google"], "alt-svc"=>["h3=\":443\"; ma=2592000,h3-29=\":443\"; ma=2592000"], "connection"=>["close"], "transfer-encoding"=>["chunked"]}
`

If the link includes rel=next:
`["<https://gw.api.it.umich.edu/um/aa/ClassroomList/v2/Classrooms?$start_index=0&$count=1000>; rel=self, <https://gw.api.it.umich.edu/um/aa/ClassroomList/v2/Classrooms?$count=1000>; rel=prev, <https://gw.api.it.umich.edu/um/aa/ClassroomList/v2/Classrooms?$start_index=1000&$count=1000>; rel=next"]`
 it means that there is a new page. 

In this case, I increase the index (start_index += count) and call the API again:
`url = URI("https://gw.api.it.umich.edu/um/aa/ClassroomList/v2/Classrooms?$start_index=#{start_index}&$count=#{count}")`

If the link doesn't have rel=next:
`["<https://gw.api.it.umich.edu/um/aa/ClassroomList/v2/Classrooms?$start_index=1000&$count=1000>; rel=self, <https://gw.api.it.umich.edu/um/aa/ClassroomList/v2/Classrooms?$count=1000>; rel=prev"]`
there is no other page, and next_page = false stops the loop that calls API.

What do you suggest to change in this code?

Reply:
I think the document is trying to point out that we can directly use the URL coming in the response instead of building it.

_Example ( Please check the link marked in bold )_

{"cache-control"=>["private"], "content-language"=>["en"], "vary"=>["accept-encoding"], "content-type"=>["application/json;subtype=denodo-8.0;charset=UTF-8"], "date"=>["Tue, 26 Mar 2024 02:41:15 GMT"], "link"=>["<https://gw.api.it.umich.edu/um/aa/ClassroomList/v2/Classrooms?$start_index=0&$count=1000>; rel=self, <https://gw.api.it.umich.edu/um/aa/ClassroomList/v2/Classrooms?$count=1000>; rel=prev, <**https://gw.api.it.umich.edu/um/aa/ClassroomList/v2/Classrooms?$start_index=1000&$count=1000**>; rel=next"], "x-request-id"=>["35c06169-f33f-4161-a4dd-aa45c00cbd5b"], "via"=>["1.1 google"], "alt-svc"=>["h3=\":443\"; ma=2592000,h3-29=\":443\"; ma=2592000"], "connection"=>["close"], "transfer-encoding"=>["chunked"]} 

Do you think this is a better approach than building the URL on our own ?

Reply:
When I call the API for the first time I use the URL with start values for start_index and count (line 100):
```
url = URI("https://gw.api.it.umich.edu/um/aa/ClassroomList/v2/Classrooms?$start_index=#{start_index}&$count=#{count}")
```
When I get a response, first I have to check if 'rel=next' is present in the returned header. If not I don't need to call the API again, if it does, I call the API with start_index += count.
I am not changing the URL variable, I only increase the start_index.
```
start_index = 0
count = 1000
next_page = true
while next_page do
  url = URI("https://gw.api.it.umich.edu/um/aa/ClassroomList/v2/Classrooms?$start_index=#{start_index}&$count=#{count}")
  ...
  if link.include? "rel=next"
    start_index += count
  else
    next_page = false
  end
  ...
end # end of loop
```

What is wrong with this approach? How do you suggest to change this code?

Reply:
I already reviewed this code and I understand the logic. I am trying to point out the usage approach mentioned  in the document.

Reference - https://dir.api.it.umich.edu/guides/migration-to-denodo

Reply:
Also, lets say the last page has only 800 results. With the current logic, we will be requesting 1000 results since we are building the URL by our own. I assume if we use the direct link approach it might have the correct value of 800 since it is returned in api response.

Reply:
I think you keep the logic as you have right now. Only change would be after "_rel=next_" condition check, consider using the direct link from response instead of building it.

Reply:
> I think you keep the logic as you have right now. Only change would be after "rel=next" condition check, consider using the direct link from response instead of building it.

Could you please write the code for geting the direct link from the response?
The returned result is a hash. The link is a string:
```
"link"=>["<https://gw.api.it.umich.edu/um/aa/ClassroomList/v2/Classrooms?$start_index=0&$count=1000>; rel=self, <https://gw.api.it.umich.edu/um/aa/ClassroomList/v2/Classrooms?$count=1000>; rel=prev, <https://gw.api.it.umich.edu/um/aa/ClassroomList/v2/Classrooms?$start_index=1000&$count=1000>; rel=next"]
```
How do you suggest getting "<https ://gw.api.it.umich.edu/um/aa/ClassroomList/v2/Classrooms?$start_index=1000&$count=1000>" from that string?
Also, with my approach, I can use any number for the count value, not only the default one (1000)

Reply:
> > I assume if we use the direct link approach it might have the correct value of 800 since it is returned in api response.
> > No, their URL has the same $count=1000. In advance, they don't know how many records the API returns.  If the return result has less than 1000 records the returned header will have no 'rel=next'.

did you confirm with them via email / chat / ticket ?

Reply:
> > I think you keep the logic as you have right now. Only change would be after "rel=next" condition check, consider using the direct link from response instead of building it.
> 
> Could you please write the code for geting the direct link from the response? The returned result is a hash. The link is a string:
> 
> ```
> "link"=>["<https://gw.api.it.umich.edu/um/aa/ClassroomList/v2/Classrooms?$start_index=0&$count=1000>; rel=self, <https://gw.api.it.umich.edu/um/aa/ClassroomList/v2/Classrooms?$count=1000>; rel=prev, <https://gw.api.it.umich.edu/um/aa/ClassroomList/v2/Classrooms?$start_index=1000&$count=1000>; rel=next"]
> ```
> 
> How do you suggest getting "<https ://gw.api.it.umich.edu/um/aa/ClassroomList/v2/Classrooms?$start_index=1000&$count=1000>" from that string? Also, with my approach, I can use any number for the count value, not only the default one (1000)

I will check